dev/run-tests

```diff
@@ -17,239 +17,394 @@
+        # remove any empty strings
+        changed_files = [f for f in raw_output.split('\n') if f]
+
+        # find any sql files
+        sql_files = [f for f in changed_files
+                     if multi_starts_with(f, 
+                                          "sql/", 
+                                          "bin/spark-sql", 
+                                          "sbin/start-thriftserver.sh")]
+
+        non_sql_files = set(changed_files).difference(set(sql_files))
+
+        if non_sql_files:
+            test_suite.append("CORE")
+        if sql_files:
+            print "[info] Detected changes in SQL. Will run Hive test suite."
+            test_suite.append("SQL")
+            if not non_sql_files:
+                print "[info] Detected no changes except in SQL. Will only run SQL tests."
+        return set(test_suite)
+    else:
+        # we aren't in the Amplab environment so merely run all tests
+        test_suite.append("CORE")
+        test_suite.append("SQL")
+        return set(test_suite)
+
+def set_title_and_block(title, err_block):
+    os.environ["CURRENT_BLOCK"] = error_codes[err_block]
+    line_str = "".join(['='] * 72)
+
+    print
+    print line_str
+    print title
+    print line_str
+
+def run_cmd(cmd):
+    """Given a command as a list of arguments will attempt to execute the
+    command and, on failure, print an error message"""
+
+    if not isinstance(cmd, list):
+        cmd = cmd.split()
+    try:
+        sp.check_output(cmd)
+    except sp.CalledProcessError as e:
+        print "[error] running", e.cmd, "; received return code", e.returncode
+        exit(e.returncode)
+
+def run_apache_rat_checks():
+    set_title_and_block("Running Apache RAT checks", "BLOCK_RAT")
+    run_cmd(["./dev/check-license"])
+
+def run_scala_style_checks():
+    set_title_and_block("Running Scala style checks", "BLOCK_SCALA_STYLE")
+    run_cmd(["./dev/lint-scala"])
+
+def run_python_style_checks():
+    set_title_and_block("Running Python style checks", "BLOCK_PYTHON_STYLE")
+    run_cmd(["./dev/lint-python"])
+
+def exec_maven(mvn_args = []):
+    """Will call Maven in the current directory with the list of mvn_args passed
+    in and returns the subprocess for any further processing"""
+
+    return sp.Popen(["./build/mvn"] + mvn_args)
+
+def exec_sbt(sbt_args = []):
+    """Will call SBT in the current directory with the list of mvn_args passed
+    in and returns the subprocess for any further processing"""
+
+    # NOTE: echo "q" is needed because sbt on encountering a build file
+    # with failure (either resolution or compilation) prompts the user for
+    # input either q, r, etc to quit or retry. This echo is there to make it
+    # not block.
+    echo_proc = sp.Popen(["echo", "\"q\n\""])
+    sbt_proc = sp.Popen(["./build/sbt"] + sbt_args,
+                        stdin=echo_proc.stdout,
+                        stdout=sp.PIPE)
+    echo_proc.wait()
```

Comment:
It's probably better here to replace this with `communicate()` since [`wait()` can be problematic](https://docs.python.org/2/library/subprocess.html#subprocess.Popen.wait). This will also let you get rid of the previous `echo_proc` line.

Perhaps something like:

```
sbt_proc = subprocess.Popen(
    ["./build/sbt"] + sbt_args,
    stdin=subprocess.PIPE)
sbt_proc.communicate('q')
```

You might not even need the logic you have just below to print the output, since I think this will output straight to stdout.